mod/redir.php

```diff
@@ -44,102 +47,105 @@ function redir_init(App $a) {
+		$remotehost = substr($contact['addr'], strpos($contact['addr'], '@') + 1);
+
+		// On a local instance we have to check if the local user has already authenticated
+		// with the local contact. Otherwise the local user would ask the local contact
+		// for authentification everytime he/she is visiting a profile page of the local
+		// contact.
+		if (($host == $remotehost) && (Session::getRemoteContactID(Session::get('visitor_visiting')) == Session::get('visitor_id'))) {
+			// Remote user is already authenticated.
+			redir_check_url($contact_url, $url);
+			$target_url = $url ?: $contact_url;
+			Logger::log($contact['name'] . " is already authenticated. Redirecting to " . $target_url, Logger::DEBUG);
+			$a->redirect($target_url);
 		}
+	}
 
-		// Doing remote auth with dfrn.
-		if (local_user() && (!empty($contact['dfrn-id']) || !empty($contact['issued-id'])) && empty($contact['pending'])) {
-			$dfrn_id = $orig_id = (($contact['issued-id']) ? $contact['issued-id'] : $contact['dfrn-id']);
+	// Doing remote auth with dfrn.
+	if (local_user() && (!empty($contact['dfrn-id']) || !empty($contact['issued-id'])) && empty($contact['pending'])) {
+		$dfrn_id = $orig_id = (($contact['issued-id']) ? $contact['issued-id'] : $contact['dfrn-id']);
 
-			if ($contact['duplex'] && $contact['issued-id']) {
-				$orig_id = $contact['issued-id'];
-				$dfrn_id = '1:' . $orig_id;
-			}
-			if ($contact['duplex'] && $contact['dfrn-id']) {
-				$orig_id = $contact['dfrn-id'];
-				$dfrn_id = '0:' . $orig_id;
-			}
+		if ($contact['duplex'] && $contact['issued-id']) {
+			$orig_id = $contact['issued-id'];
+			$dfrn_id = '1:' . $orig_id;
+		}
+		if ($contact['duplex'] && $contact['dfrn-id']) {
+			$orig_id = $contact['dfrn-id'];
+			$dfrn_id = '0:' . $orig_id;
+		}
 
-			$sec = Strings::getRandomHex();
+		$sec = Strings::getRandomHex();
 
-			$fields = ['uid' => local_user(), 'cid' => $cid, 'dfrn_id' => $dfrn_id,
-				'sec' => $sec, 'expire' => time() + 45];
-			DBA::insert('profile_check', $fields);
+		$fields = ['uid' => local_user(), 'cid' => $cid, 'dfrn_id' => $dfrn_id,
+			'sec' => $sec, 'expire' => time() + 45];
+		DBA::insert('profile_check', $fields);
 
-			Logger::log('mod_redir: ' . $contact['name'] . ' ' . $sec, Logger::DEBUG);
+		Logger::log('mod_redir: ' . $contact['name'] . ' ' . $sec, Logger::DEBUG);
 
-			$dest = (!empty($url) ? '&destination_url=' . $url : '');
+		$dest = (!empty($url) ? '&destination_url=' . $url : '');
 
-			System::externalRedirect($contact['poll'] . '?dfrn_id=' . $dfrn_id
-				. '&dfrn_version=' . DFRN_PROTOCOL_VERSION . '&type=profile&sec=' . $sec . $dest . $quiet);
-		}
+		System::externalRedirect($contact['poll'] . '?dfrn_id=' . $dfrn_id
+			. '&dfrn_version=' . DFRN_PROTOCOL_VERSION . '&type=profile&sec=' . $sec . $dest . $quiet);
 
+		redir_check_url($contact_url, $url);
```

Comment:
`System::externalRedirect` exits or throws an exception, nothing after it is executed.

Reply:
?